test/lit/basic/shared-types.wast

```diff
@@ -0,0 +1,45 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.
+
+;; RUN: wasm-opt %s -all -S -o - | filecheck %s
+;; RUN: wasm-opt %s -all --roundtrip -S -o - | filecheck %s
+
+(module
+  ;; CHECK:      (type $func (shared (func)))
+
+  ;; CHECK:      (type $1 (func))
+
+  ;; CHECK:      (type $final (shared (struct )))
+  (type $final (shared (struct)))
+  ;; CHECK:      (type $top (sub (shared (struct ))))
+  (type $top (sub (shared (struct))))
+  ;; CHECK:      (type $mid (sub $top (shared (struct (field i32)))))
+  (type $mid (sub $top (shared (struct i32))))
+  ;; CHECK:      (type $bot (sub final $mid (shared (struct (field i32) (field i32)))))
+  (type $bot (sub final $mid (shared (struct i32 i32))))
+
+  (type $func (shared (func)))
```

Comment:
Can move this above to the CHECK.

Reply:
I'll use a rec group to get more control over the ordering.